library/utf8/src/commonMain/kotlin/io/matthewnelson/encoding/utf8/UTF8.kt

```diff
@@ -0,0 +1,469 @@
+                    out.output('?'.code.toByte())
+                }
+                override fun toString(): String = "UTF8.ReplacementStrategy[U+0034]"
+            }
+
+            /**
+             * TODO
+             * */
+            @JvmField
+            public val U_FFFD: ReplacementStrategy = object : ReplacementStrategy(size = 3) {
+                override fun doOutput(out: Decoder.OutFeed) {
+                    out.output(0xef.toByte())
+                    out.output(0xbf.toByte())
+                    out.output(0xbd.toByte())
+                }
+                override fun toString(): String = "UTF8.ReplacementStrategy[U+FFFD]"
+            }
+
+            /**
+             * TODO
+             * */
+            @JvmField
+            public val KOTLIN: ReplacementStrategy = initKotlin(U_0034, U_FFFD)
+
+            /**
+             * TODO
+             * */
+            @JvmField
+            public val THROW: ReplacementStrategy = object : ReplacementStrategy(size = 0) {
+                override fun doOutput(out: Decoder.OutFeed) {
+                    throw EncodingException("Malformed UTF-8 character sequence")
+                }
+                override fun sizeOrThrow(): Int {
+                    throw EncodingException("Malformed UTF-8 character sequence")
+                }
+                override fun toString(): String = "UTF8.ReplacementStrategy[THROW]"
+            }
+        }
+
+        @Throws(EncodingException::class)
+        internal abstract fun doOutput(out: Decoder.OutFeed)
+        @Throws(EncodingException::class)
+        internal open fun sizeOrThrow(): Int = size
+
+        /** @suppress */
+        public abstract override fun toString(): String
+    }
+
+    /**
+     * TODO
+     * */
+    public class Config private constructor(
+
+        /**
+         * TODO
+         * */
+        @JvmField
+        public val alwaysUsePreProcessing: Boolean,
+
+        /**
+         * TODO
+         * */
+        @JvmField
+        public val replacementStrategy: ReplacementStrategy,
+    ): EncoderDecoder.Config(isLenient = null, lineBreakInterval = -1, paddingChar = null) {
+
+        protected override fun decodeOutMaxSizeProtected(
+            encodedSize: Long,
+        ): Long = if (encodedSize > MAX_CHAR_LEN) Long.MIN_VALUE else encodedSize * 3
+
+        protected override fun decodeOutMaxSizeOrFailProtected(
+            encodedSize: Int,
+            input: DecoderInput,
+        ): Int {
+            // TODO: Annotate EncoderDecoder.Config.encodeOutSize with new exception
+            run {
+                if (alwaysUsePreProcessing) return@run
+                val s = decodeOutMaxSizeProtected(encodedSize.toLong())
+                if (s in 0L..Int.MAX_VALUE.toLong()) return s.toInt()
+                // TODO: If (constantTime) return Int.MIN_VALUE
+                // Try pre-processing to see if it can fit.
+            }
+            val pp = CharPreProcessor(replacementStrategy)
```

Comment:
If the input is large enough, this can take several seconds on Native/Js. Removing the option would be best imo.